Add filtering by kind and sorting by name or kind to the Unified Resources view

bl-nero opened the pull request:
Note that the changes don't replicate the UI design perfectly, but they are the best compromise between the speed of delivery and the actual functionality. They currently reuse existing components.

Screenshot:
![image](https://github.com/gravitational/teleport/assets/9391503/22eb1730-730e-4236-8182-9300aca9e5ac)

Related issue: #27462

avatus committed:
Update resourceKind method to return slice

avatus committed:
Adds a UnifiedResource watcher to the auth server. This watcher
will watch all the types that are displayed in the web UI and store
them in-memory to allow us to search/filter/query and get multiple
kinds returned at the same time.
[RFD](https://github.com/gravitational/teleport/pull/28162)

avatus committed:
Add UnifiedResourceComponent

avatus committed:
Add unifiedResourceWatcher to auth server

avatus committed:
Init unified resource watcher

avatus committed:
Convert to strings for error output

avatus committed:
Add license

bl-nero committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

bl-nero committed:
Add some tests

bl-nero committed:
Merge remote-tracking branch 'origin/master' into michaelmyers/unified_resource_watcher

bl-nero committed:
Fix a race condition and properly close the watcher

avatus committed:
add defer

avatus committed:
Add kubes to watcher

avatus committed:
Add uninit check

avatus committed:
Notify if stale

avatus committed:
Add SAML to watcher

avatus committed:
Update test and fix lint

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Fix license

avatus committed:
Add godocs

avatus committed:
Add Kind to ui structs

avatus committed:
Add Kind to ui structs

avatus committed:
Add FakePaginateParams

avatus committed:
Add FakePaginateParams

avatus committed:
Update ListUnifiedResources proto messages

avatus committed:
Update ListUnifiedResources proto messages

avatus committed:
Update uses of FakePaginate

avatus committed:
Update uses of FakePaginate

avatus committed:
Include type in seenmap for MatchResourceByFilters

avatus committed:
Include type in seenmap for MatchResourceByFilters

avatus committed:
Update TestAuthServer with unified resource watcher

avatus committed:
Update TestAuthServer with unified resource watcher

avatus committed:
CheckAndSetDefaults for ListUnifiedResourcesRequest

avatus committed:
CheckAndSetDefaults for ListUnifiedResourcesRequest

avatus committed:
Add ListUnifiedResources rpc

avatus committed:
Add ListUnifiedResources rpc

avatus committed:
Update protos

avatus committed:
Update protos

avatus committed:
Add test

avatus committed:
Add test

avatus committed:
Add docs and update tests

avatus committed:
Add docs and update tests

avatus committed:
Add more docs

avatus committed:
Add more docs

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Combine app types in switch

avatus committed:
Combine app types in switch

avatus committed:
Fix tests

avatus committed:
Fix tests

avatus committed:
Add kube access checker

avatus committed:
Add kube access checker

avatus committed:
Add remaining kinds

avatus committed:
Add remaining kinds

bl-nero committed:
Add a Unified Resource card component

bl-nero committed:
WIP: Unified Resource list

bl-nero committed:
Lint fixes

avatus committed:
Review feedback

avatus committed:
Add configurable queuesize to NewWatcher

avatus committed:
Move Get to collector

bl-nero committed:
Review comments

bl-nero committed:
Rename AgentKind and fix lint issues

avatus committed:
make unified resource cache public

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

avatus committed:
Move to common interface

avatus committed:
Use copy as value

avatus committed:
Move saml getter

bl-nero committed:
Review comments

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
WIP: infinite scrolling

bl-nero committed:
Prevent VS Code from showing errors in tsconfig

If tsconfig doesn't have an `outDir` option set, VS Code shows multitude of errors of the following kind:

```
Cannot write file '/Users/bartosz/code/teleport/e/web/teleport/babel.config.js' because it would overwrite input file.
```

bl-nero committed:
Remove unnecessary files

bl-nero committed:
todo

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Cleanup

bl-nero committed:
Add license

avatus committed:
invert watcher and cache

avatus committed:
use cache instead of watcher

avatus committed:
Fix imports

avatus committed:
Fix lint: check for err

avatus committed:
review round 1

avatus committed:
Update test

bl-nero committed:
Merge branch 'michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Review, fixing broken tests

avatus committed:
Add copy to get

bl-nero committed:
Add a search panel to the unified resources list

avatus committed:
Add CloneAny interface to help with unified resource cloning

avatus committed:
Use cloneany

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Add sort for name and kind

avatus committed:
Add app and saml, fix tests

bl-nero committed:
Review

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Review

bl-nero committed:
Lint

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

avatus committed:
Watch kube server instead of cluster

avatus committed:
Watch kube server instead of cluster

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Get kubecluster from server

avatus committed:
merge master

avatus committed:
update proto file

avatus committed:
Add no access test case

bl-nero committed:
Review, fix CI problems

bl-nero committed:
One more lint fix

bl-nero committed:
Yet another lint fix

avatus committed:
move sorting

avatus committed:
Fix sorting by name

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

avatus committed:
Review feedback round 2

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Remove unneeded bk return

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Keep grpc naming convention

avatus committed:
Add missing docs

avatus committed:
defer log

bl-nero committed:
WIP: filter and sort

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

bl-nero committed:
Make sorting work

bl-nero committed:
Make filters (almost) work

bl-nero committed:
Fix the filters

bl-nero committed:
Fix filter history handling and kind-based sorting

bl-nero committed:
Merge branch 'michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

bl-nero committed:
Visual tweaks

bl-nero committed:
lint and test fixes

bl-nero committed:
Add licenses

zmb3 commented:
Can you include a screenshot or description of where we deviated from the design? It's nice to have that in the PR for historical reasons.

avatus commented on lib/auth/auth_with_roles.go:
```diff
@@ -1717,18 +1690,8 @@ func (a *ServerWithRoles) ListUnifiedResources(ctx context.Context, req *proto.L
-		default:
-			return nil, trace.NotImplemented("sorting by field %q for unified resource %q is not supported", req.SortBy.Field, types.KindUnifiedResource)
+		if err := filteredResources.SortByCustom(req.SortBy); err != nil {
+			return nil, err
```
```suggestion
			return nil, trace.Wrap(err, "sorting unified resources")
```

avatus replied:
```suggestion
			return nil, trace.Wrap(err, "sorting unified resources")
```

zmb3 commented on e:
Unintentional e update?

avatus commented on web/packages/teleport/src/UnifiedResources/FilterPanel.tsx:
```diff
@@ -0,0 +1,156 @@
+}
+
+function oppositeSort(sort: SortType): SortType {
+  switch (sort.dir) {
+    case 'ASC':
+      return { ...sort, dir: 'DESC' };
+    case 'DESC':
+      return { ...sort, dir: 'ASC' };
+    default:
+      // Will never happen. Of course.
+      return sort;
+  }
```
this is fine. we could also just check for `DESC` and make the default the `ASC` case

bl-nero replied:
I somehow prefer to be a bit more explicit :)

avatus commented on api/types/constants.go:
```diff
@@ -952,6 +952,9 @@ const (
 	ResourceSpecType = "type"
+
+	// ResourceSpecKind refers to a resource field named "kind".
+	ResourceSpecKind = "kind"
```
I don't believe `Kind` is part of a resources's spec. Maybe just change to `ResourceKind`?

bl-nero committed:
Review

bl-nero committed:
Fix the useEffect dependency list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

avatus committed:
Review feedback

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

avatus reviewed (approved):
If possible without too much change, I'd like to get the search bar/filter styles in line with the desgn (just taller/thicker and whatever). but if it requires many external file changes, then we can push this in and separate that out once its merged.

bl-nero commented:
I'd rather keep intact those already reviewed changes that bring us to the
MVP state, and then add more stuff on top of it in separate PRs.

On Thu, Aug 10, 2023 at 9:02 PM Michael ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> If possible without too much change, I'd like to get the search bar/filter
> styles in line with the desgn (just taller/thicker and whatever). but if it
> requires many external file changes, then we can push this in and separate
> that out once its merged.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitational/teleport/pull/30124#pullrequestreview-1572556636>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHU3D6ZI6TGRZKOJ7VOAP3XUUV5HANCNFSM6AAAAAA3G6T2UU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

avatus committed:
Add ListUnifiedResources gRPC and web endpoints (#29661)

Adds the `ListUnifiedResources` grpc endpoint that returns paginated unified resources and exposes it in the web apiserver

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_resource_watcher' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

bl-nero committed:
Merge branch 'master' into bl-nero/unified-resources-sort

bl-nero committed:
Merge branch 'master' into bl-nero/unified-resources-sort

bl-nero committed:
Fix tests
